packages/ckeditor5-table/src/tableselection/croptable.js

```diff
@@ -7,146 +7,168 @@
+		const row = croppedTable.getChild( rowInCroppedTable );
 
-	for ( const tableCell of selectedTableCells ) {
-		const row = findAncestor( 'tableRow', tableCell );
+		// For empty slots: fill the gap with empty table cell.
+		if ( isSpanned ) {
+			// TODO: Remove table utils usage. See: https://github.com/ckeditor/ckeditor5/issues/6785.
+			const { row: anchorRow, column: anchorColumn } = tableUtils.getCellLocation( tableCell );
 
-		if ( !rowToCopyMap.has( row ) ) {
-			const rowCopy = row._clone();
-			writer.append( rowCopy, tableCopy );
-			rowToCopyMap.set( row, rowCopy );
+			// But fill the gap only if the spanning cell is anchored outside cropped area.
+			// In the table from method jsdoc those cells are: "c" & "f".
+			if ( anchorRow < startRow || anchorColumn < startColumn ) {
+				createEmptyTableCell( writer, writer.createPositionAt( row, 'end' ) );
+			}
 		}
+		// Otherwise clone the cell with all children and trim if it exceeds cropped area.
+		else {
+			const tableCellCopy = tableCell._clone( true );
 
-		const tableCellCopy = tableCell._clone( true );
-		const { column } = tableUtils.getCellLocation( tableCell );
+			writer.append( tableCellCopy, row );
 
-		copyToOriginalColumnMap.set( tableCellCopy, column );
-
-		writer.append( tableCellCopy, rowToCopyMap.get( row ) );
+			// Trim table if it exceeds cropped area.
+			// In the table from method jsdoc those cells are: "g" & "m".
+			trimTableCellIfNeeded( tableCellCopy, sourceRow, sourceColumn, endRow, endColumn, writer );
+		}
 	}
 
-	addMissingTableCells( tableCopy, startColumn, copyToOriginalColumnMap, writer, tableUtils );
+	// Adjust heading rows & columns in cropped table if crop selection includes headings parts.
+	addHeadingsToCroppedTable( croppedTable, sourceTable, startRow, startColumn, writer );
 
-	return tableCopy;
+	return croppedTable;
 }
 
-// Fills gaps for spanned cell from outside the selection range.
-function addMissingTableCells( tableCopy, startColumn, copyToOriginalColumnMap, writer, tableUtils ) {
-	for ( const row of tableCopy.getChildren() ) {
-		for ( const tableCell of Array.from( row.getChildren() ) ) {
-			const { column } = tableUtils.getCellLocation( tableCell );
+/**
+ * Returns a cropped table from the selected table cells.
+ *
+ * This function is to be used with the table selection.
+ *
+ *		tableSelection.setCellSelection( startCell, endCell );
+ *
+ *		const croppedTable = cropTable( tableSelection.getSelectedTableCells(), tableUtils, writer );
+ *
+ * **Note**: This function is also used by {@link module:table/tableselection~TableSelection#getSelectionAsFragment}.
+ *
+ * @param {Iterable.<module:engine/model/element~Element>} selectedTableCellsIterator
+ * @param {module:engine/model/writer~Writer} writer
+ * @param {module:table/tableutils~TableUtils} tableUtils
+ * @returns {module:engine/model/element~Element}
+ */
+export function cropTableToSelection( selectedTableCellsIterator, writer, tableUtils ) {
```

Comment:
I'd move this util to TableSelection as it's only used there and it's a bit hard to document.